### PR TITLE
Feature/bonding curve

### LINKED ISSUE #819: Bonding Curve aka Automatic Market Maker Contract
## 🎉 Description

I'm working on a PR to add a simple bonding curve contract (aka automatic market maker) based on Bancor formula. Here is the reference contract: https://github.com/relevant-community/contracts/blob/master/contracts/BondingCurve.sol

I have a few questions about structure and dependencies:
- [x] Folder structure —where should this contract go? ```/token```? a separate ```/bondingcurve``` folder?  

Planning on adding a ```curation-markets``` folder width ```BancorFormula.sol``` and ```BondingCurve.sol```. Adding ```Power.sol``` function to the ```math``` folder.
- [x] The contract depends on the ```BancorFormula.sol``` contract with some fancy tricks to compute fraction exponents and comes along with python folder to generate constants and test them: https://github.com/bancorprotocol/contracts/tree/master/solidity/python Do we want to include this in the zeppelin repo or is it enough to link out to it?

Extracting the power function into a  separate ```Power.sol``` (licensed under Apache 2.0) file in the ```/math``` folder.

- [x] Bancor has a .js test for BancorFormula.sol, so I assume we will want to include it: https://github.com/bancorprotocol/contracts/blob/master/solidity/test/BancorFormula.js   

Renaming this to ```Power.test.js```. The file is very lengthy — its iterating through all the constants generated via python — do we want to keep it?

- [x] license - Bancor code is licensed under Apache License v2.0 and OpenZeppelin is MIT, is there a conflict here? do we need some special notice when we use Bancor code?  

Adding a an Apache 2.0 License and modification notice to the  ```BancorFormula.sol``` and ```Power.sol``` files with links to the Bancor repo. 

-------

- [ ] 🐛 This is a bug report.
- [x] 📈 This is a feature request.

## 👍 Other Information

Some background reading about bonding curves:  
https://hackernoon.com/how-to-make-bonding-curves-for-continuous-token-models-3784653f8b17
https://medium.com/@simondlr/tokens-2-0-curved-token-bonding-in-curation-markets-1764a2e0bee5
and Bancor formula:
https://drive.google.com/file/d/0B3HPNP-GDn7aRkVaV3dkVl9NS2M/view

### PR DESCRIPTION
**Pull request by balasan**
Fixes #819

# 🚀 Description

Adding a simple bonding curve contract (aka automatic market maker) based on Bancor formula. The codes is based on and includes code by Bancor: https://github.com/bancorprotocol/contracts. 

I have extracted the ```power``` function from the original ```BancorFormula.sol``` contract into a separate ```Power.sol``` contract in the ```math``` folder. This function depends on some pre-generated constants and is beyond my understanding. Although there is a test for it, including it in the library does depend on trusting the Bancor code to some degree and might be hard to audit. I can see this being a reason to not merge this PR, but differing the decision to the Zeppelin team. 

I have also pushed the code to a separate repo: https://github.com/relevant-community/bonding-curve so in the case it is not a good fit for Zeppelin we can maintain it there.

### Some background reading about bonding curves:
https://hackernoon.com/how-to-make-bonding-curves-for-continuous-token-models-3784653f8b17
https://medium.com/@simondlr/tokens-2-0-curved-token-bonding-in-curation-markets-1764a2e0bee5
and Bancor formula:
https://drive.google.com/file/d/0B3HPNP-GDn7aRkVaV3dkVl9NS2M/view

- [x] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](/docs/CONTRIBUTING.md)
- [x] ✅ I've added tests where applicable to test my new functionality.
- [x] 📖 I've made sure that my contracts are well-documented.
- [x] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:all:fix`).

### CODE REVIEW

**Comment by Shrugs** [comment]
@balasan 

there are a few linting errors on Power.sol, primarily along the usage of the `[`. run `npm run lint:sol` to see all of them or check the output at https://travis-ci.org/OpenZeppelin/zeppelin-solidity/jobs/356012100

The error was caused by a ganache error, restarting the job now

File: contracts/curation-markets/BancorFormula.sol [hunk 1]
```diff
@@ -0,0 +1,106 @@
+pragma solidity ^0.4.18;
+import "../math/Power.sol";
+import "../math/SafeMath.sol";
+
+
+/**
+ * Bancor formula by Bancor
+ * https://github.com/bancorprotocol/contracts
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: `@dev` prefixes

File: contracts/curation-markets/BancorFormula.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    uint256 _supply,
+    uint256 _connectorBalance,
+    uint32 _connectorWeight,
+    uint256 _depositAmount) public constant returns (uint256)
+  {
+    // validate input
+    require(_supply > 0 && _connectorBalance > 0 && _connectorWeight > 0 && _connectorWeight <= MAX_WEIGHT);
+
+    // special case for 0 deposit amount
+    if (_depositAmount == 0)
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: `{ }` to match codebase style

File: contracts/curation-markets/BancorFormula.sol [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      return 0;
+
+    // special case if the weight = 100%
+    if (_connectorWeight == MAX_WEIGHT)
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: likewise `{ }` here

File: contracts/curation-markets/BancorFormula.sol [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    uint256 _supply,
+    uint256 _connectorBalance,
+    uint32 _connectorWeight,
+    uint256 _sellAmount) public constant returns (uint256)
+  {
+    // validate input
+    require(_supply > 0 && _connectorBalance > 0 && _connectorWeight > 0 && _connectorWeight <= MAX_WEIGHT && _sellAmount <= _supply);
+
+    // special case for 0 sell amount
+    if (_sellAmount == 0)
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: `{ }` here

File: contracts/curation-markets/BancorFormula.sol [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      return 0;
+
+    // special case for selling the entire supply
+    if (_sellAmount == _supply)
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: here

File: contracts/curation-markets/BancorFormula.sol [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      return _connectorBalance;
+
+    // special case if the weight = 100%
+    if (_connectorWeight == MAX_WEIGHT)
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: here

File: contracts/curation-markets/BancorFormula.sol [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      return _connectorBalance.mul(_sellAmount).div(_supply);
+
+    uint256 result;
+    uint8 precision;
+    uint256 baseD = _supply - _sellAmount;
+    (result, precision) = power(
+      _supply, baseD, MAX_WEIGHT, _connectorWeight
+    );
+    uint256 temp1 = _connectorBalance.mul(result);
```
Comment (Shrugs) [inline contracts/curation-markets/BancorFormula.sol]: is there a better name for `temp1` and `temp2`?

File: contracts/curation-markets/BondingCurve.sol [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * @title Bonding Curve
+ * @dev Bonding curve contract based on Bacor formula
+ * inspired by bancor protocol and simondlr
+ * https://github.com/bancorprotocol/contracts
+ * https://github.com/ConsenSys/curationmarkets/blob/master/CurationMarkets.sol
+ */
+contract BondingCurve is StandardToken, BancorFormula, Ownable {
+  uint256 public poolBalance;
```
Comment (Shrugs) [inline contracts/curation-markets/BondingCurve.sol]: documentation for this variable?

File: contracts/curation-markets/BondingCurve.sol [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+  /*
+    reserve ratio, represented in ppm, 1-1000000
```
Comment (Shrugs) [inline contracts/curation-markets/BondingCurve.sol]: Can you use the 

```
/**
 * asfsd
 * asdf
 */  
```

style for these block comments as we do elsewhere?

Comment (Shrugs) [reply contracts/curation-markets/BondingCurve.sol]: they should probably also be prefixed with `@dev`

**Comment by balasan** [comment]
seems like ```npm run lint:sol:fix``` edited a few other files hope thats fine...

**Comment by Shrugs** [comment]
LGTM, but I'd like an expert in BondingCurve theory to take a peek as well. Can Billy come review this as well? cc @spalladino as well, since I know you did some research into TCRs for zeppelin_os

**Comment by okwme** [comment]
* make BondingCurveMock.sol constructor payable so that the initial `_poolBalance` can be filled without triggering the fallback `buy()` function
* make BondingCurveMock.sol constructor assign the `totalSupply_` to an address so that they actually exist
* truffle-config.js needs to use a gasPrice that matches the gasPrice in `BondingCurve.test.js` (might only be relevant to relevant-community/bonding-curve)

**Comment by balasan** [comment]
Thanks @okwme commited all of those updates
One new issue I found is that the contract becomes stuck if everyone sells all tokens and totalSupply & poolBalance are 0. This is because ```BancorFormula``` does not work when either of those values is 0. 
Issue + proposed solutions opened here: https://github.com/relevant-community/bonding-curve/issues/2

**Comment by oed** [comment]
@balasan Cool stuff! Need to dive deeper into the code :D 
Q: do you have any plans for supporting other tokens than just ether as the bonding token?

**Comment by barakman** [comment]
@balasan:
BancorFormula writer here (the contract, not the formula).

With regards to "This function depends on some pre-generated constants and is beyond my understanding" - I can explain into details if you're interested.
In any case, each set of constants can be generated by a script in Python (as documented in the contract).
All of these constants are generally dependent on two constants only - `MAX_PRECISION` (set to 127) and the number of coefficients in the exponentiation function (set to 34).
Anyone using this code may choose to play around with those two constants, which would ultimately impact accuracy and performance (which are "tightly coupled" in a typical "trade-off relationship").
Initially, the code was implemented with a fixed `PRECISION` of 32. With this level of precision, it was imperative to use 34 coefficients in order to ensure reasonable accuracy. However, after introducing dynamic precision (which can go as high as 127), the number of coefficients becomes much less critical for the sake of accuracy. For a given input, using a smaller number of coefficients would allow a higher value of `precision` passed to the exponentiation function (while improving performance and reducing gas cost).

BTW, this contract has since been optimized (for performance mostly), and you can find it at https://github.com/bancorprotocol/contracts/tree/0.3.18 (though the implementation is slightly more cumbersome, so I doubt you would want to incorporate it into Zeppelin).

With regards to testing - there are `web3` tests in both Python in Javascript.
But while the JS test is primarily for sanity assertion, the Python testing infrastructure provides a much wider coverage (regression if you will). I can go into details here too if needed.

With regards to "the contract becomes stuck if everyone sells all tokens and totalSupply & poolBalance are 0" - the same thing would happen if everyone buys until supply/balance are `MAX_NUM` (2 ^ 129) or higher. I'm not so sure that this is a plausible scenario however (nor is the scenario that you've described), but that's probably a product-related decision.

Thanks

**Comment by nventuro** [comment]
Hello @balasan, and thank you for your interest and contribution! I'm sorry so much time has passed with no response from us.

I really want for Bonding Curves to be added to OZ and want to get this merged, but worry that the PR as it is has become too large (there's the Bonding Curve itself, the Bancor formula, the Power library, tests, etc.).

Would you be willing to split this into smaller, self-contained PRs so that we can more rapidly review changes and merge? Thanks!

**Comment by balasan** [comment]
@nventuro I think it would make sense to break out the Power function into a separate PR. Bonding Curve and Bancor formula should probably go together — does that work?

**Comment by nventuro** [comment]
@balasan since the Bancor formula is its own contract, I think it makes sense for there for be a PR introducing it, with tests, etc., and then build on top of it (same with the power functions).

**Comment by nventuro** [comment]
Closing due to staleness.

**Pull request closed without merging**